pkg/store/proxy.go

```diff
@@ -159,6 +172,8 @@ func (s *ProxyStore) Series(r *storepb.SeriesRequest, srv storepb.Store_SeriesSe
 
 	for resp := range respCh {
 		if err := srv.Send(resp); err != nil {
+			recvCancel()
+			for range respCh {}
```

Comment:
Sorry I don't get your question, but I can explain why we are not just returning error here.

So this loop is the only reader. The code before (and mostly startStreamSeriesSet) are multiple writers. As you said this is responsibility of sender to close the channel. In this case though we clearly cannot read more from channel, because reader Target endpoint (srv.Send) broke (stream back to client). This means that we need to *notify* the senders to abort their work (by canceling their context) and close the respCh. However if we would just notify them and return, sender could be just about to send msg on channel that no one reads. Since it has limited buffer it can be full, so blocking multiple go routines forever.  That's why we have empty loop that reads and discards the messages. Keep in mind that send on close channel blocks forever. 

Now when I think of it... I could just return err as you propose and fix senders to send in select with send and check if context is cancelled (: will try it!

Reply:
Sorry I don't get your question, but I can explain why we are not just returning error here.

So this loop is the only reader. The code before (and mostly startStreamSeriesSet) are multiple writers. As you said this is responsibility of sender to close the channel. In this case though we clearly cannot read more from channel, because reader Target endpoint (srv.Send) broke (stream back to client). This means that we need to *notify* the senders to abort their work (by canceling their context) and close the respCh. However if we would just notify them and return, sender could be just about to send msg on channel that no one reads. Since it has limited buffer it can be full, so blocking multiple go routines forever.  That's why we have empty loop that reads and discards the messages. Keep in mind that send on close channel blocks forever. 

Now when I think of it... I could just return err as you propose and fix senders to send in select with send and check if context is cancelled (: will try it!